### PR TITLE
Register independent

### PR DESCRIPTION
**Pull request by faucomte97**

### CODE REVIEW

**Review by SKairinos — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317)* status: all files reviewed, 5 unresolved discussions (waiting on @faucomte97)

___
*[`backend/api/serializers/user.py` line 71 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzRJsM4Nh9q6ou__Bw:-NtzRJsM4Nh9q6ou__Bx:bahd1r) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L71)):*
> ```Python
>             if user_type == "teacher":
>                 Teacher(new_user=user)
>             elif user_type == "student":
> ```

I notice in the UserViewSet.get_serializer_context you are setting user_type="independent".

That is correct because we are creating a new user (user is None in this check).

However, you are not handling it here in the serializer. should be:

```py
user_type: str = self.context["user_type"]
if user_type == "teacher":
    Teacher(new_user=user)
elif user_type == "student":
    Student(new_user=user, class_field=Class())
elif user_type == "independent":
    Student(new_user=user)
```

Furthermore, I noticed that the password validators we have in py package are slightly incorrect. Explained in py package PR.
___
*[`backend/api/serializers/user.py` line 104 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzKtQP9tjN3WFAE5XA:-NtzKtQP9tjN3WFAE5XB:b1se4af) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L104)):*
> ```Python
>         required=False,
>     )
>     date_of_birth = serializers.DateField(required=False)
> ```

set write_only=True - we do not need to return this info in the response
remove required=True - it is required when creating a indy user.
___
*[`backend/api/serializers/user.py` line 104 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzLgQ4CEN78AEtfZ_Z:-NtzLgQ4CEN78AEtfZ__:b-3x3oe7) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L104)):*
> ```Python
>         required=False,
>     )
>     date_of_birth = serializers.DateField(required=False)
> ```

need to create validate_date_of_birth(). If self.instance is not None: raise validation error "Cannot update this field" (it's a create-only field)
need to create validate_date_of_add_to_newsletter(). If self.instance is not None: raise validation error "Cannot update this field" (it's a create-only field)

Actually, because we we need create-only fields, I think it's better we have separate serializers:
1. CreateUserSerializer(BaseUserSerializer[IndependentUser]): ...
2. UpdateUserSerializer(BaseUserSerializer[User]): ...

Can we do some pair programming on this? I have an idea of how we can compartmentalise all this and make it cleaner.
___
*[`backend/api/serializers/user.py` line 181 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzQ78yCtr_mJVocrp6:-NtzQ78yCtr_mJVocrp7:b-3omv68) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L181)):*
> ```Python
> 
>     def create(self, validated_data):
>         user_fields = t.cast(DataDict, validated_data)
> ```

no need to cast to DataDict. validated_data is already of type DataDict. See here:
https://github.com/ocadotechnology/codeforlife-package-python/blob/8f26c149537cc57212d97b5972090c4c2c06447b/codeforlife/serializers/model.py#L53
___
*[`backend/api/serializers/user.py` line 182 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzQUXPCyMlc_IdACyA:-NtzQUXPCyMlc_IdACyB:b9b83di) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L182)):*
> ```Python
>     def create(self, validated_data):
>         user_fields = t.cast(DataDict, validated_data)
>         add_to_newsletter = user_fields.pop("add_to_newsletter")
> ```

need to type hint that add_to_newsletter is of type bool

**Review by faucomte97 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317)* status: 1 of 9 files reviewed, 5 unresolved discussions (waiting on @faucomte97 and @SKairinos)

___
*[`backend/api/serializers/user.py` line 71 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzRJsM4Nh9q6ou__Bw:-Nu45dW06KoawbC8LJdf:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L71)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

I notice in the UserViewSet.get_serializer_context you are setting user_type="independent".

That is correct because we are creating a new user (user is None in this check).

However, you are not handling it here in the serializer. should be:

```py
user_type: str = self.context["user_type"]
if user_type == "teacher":
    Teacher(new_user=user)
elif user_type == "student":
    Student(new_user=user, class_field=Class())
elif user_type == "independent":
    Student(new_user=user)
```

Furthermore, I noticed that the password validators we have in py package are slightly incorrect. Explained in py package PR.
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 104 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzKtQP9tjN3WFAE5XA:-Nu4Q32d2hkMrX79eQ9A:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L104)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

set write_only=True - we do not need to return this info in the response
remove required=True - it is required when creating a indy user.
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 104 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzLgQ4CEN78AEtfZ_Z:-Nu4Q3Jw5-QEcEax_ClP:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L104)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

need to create validate_date_of_birth(). If self.instance is not None: raise validation error "Cannot update this field" (it's a create-only field)
need to create validate_date_of_add_to_newsletter(). If self.instance is not None: raise validation error "Cannot update this field" (it's a create-only field)

Actually, because we we need create-only fields, I think it's better we have separate serializers:
1. CreateUserSerializer(BaseUserSerializer[IndependentUser]): ...
2. UpdateUserSerializer(BaseUserSerializer[User]): ...

Can we do some pair programming on this? I have an idea of how we can compartmentalise all this and make it cleaner.
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 181 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzQ78yCtr_mJVocrp6:-Nu4H77S9E4nCiyANWs4:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L181)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

no need to cast to DataDict. validated_data is already of type DataDict. See here:
https://github.com/ocadotechnology/codeforlife-package-python/blob/8f26c149537cc57212d97b5972090c4c2c06447b/codeforlife/serializers/model.py#L53
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 182 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzQUXPCyMlc_IdACyA:-Nu4H7kAAjUsq8hd3N_B:b-s0hjid) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L182)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

need to type hint that add_to_newsletter is of type bool
</blockquote></details>

Why? Should I do the same in the CreateTeacherSerializer?

**Review by SKairinos — CHANGES_REQUESTED** [review]
> Reviewed 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317)* status: all files reviewed, 15 unresolved discussions (waiting on @faucomte97)

___
*[`backend/Pipfile` line 48 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4QlqXCrRkfd6VA2BZ:-Nu4QlqXCrRkfd6VA2B_:b-ye0qbb) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/Pipfile#L48)):*
> ```TOML
> google-cloud-container = "==2.3.0"
> # "django-anymail[amazon_ses]" = "==7.0.*"
> djangorestframework = "*"
> ```

not needed. [inherited from CFL package](https://github.com/ocadotechnology/codeforlife-package-python/blob/fc40e3ac7c2b1d7b932e4bdd01a656846cc985a7/Pipfile#L8)
___
*[`backend/api/serializers/user.py` line 182 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzQUXPCyMlc_IdACyA:-Nu4WRZND6JH3KiQ-H1o:bgu0i1m) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L182)):*
<details><summary><i>Previously, faucomte97 (Florian Aucomte) wrote…</i></summary><blockquote>

Why? Should I do the same in the CreateTeacherSerializer?
</blockquote></details>

Because the data we pop from validated_data could be of any type. Therefore, `type(add_to_newsletter) == Any`. We need to hint to mypy that its a bool `add_to_newsletter: bool =  user_fields.pop("add_to_newsletter")`.

Yes, it should also be in CreateTeacherSerializer (that's my bad)
___
*[`backend/api/serializers/user.py` line 97 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4UiMe7jbFn-FMiJoe:-Nu4UiMe7jbFn-FMiJof:b-o7z4vd) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L97)):*
> ```Python
> 
> class CreateUserSerializer(BaseUserSerializer[IndependentUser]):
>     date_of_birth = serializers.DateField(required=False)
> ```

this is required when creating. replace with write_only=True as we do not need to return this value in the response.
___
*[`backend/api/serializers/user.py` line 128 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4VRtAA4QG002STuGs:-Nu4VRtAA4QG002STuGt:b1qf9kj) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L128)):*
> ```Python
>     requesting_to_join_class = serializers.CharField(
>         source="new_student.pending_class_request",
>         required=False,
> ```

can remove required=False. Partial update forces everything to be required=False. No need to manually specify it.
___
*[`backend/api/serializers/user.py` line 133 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4VbYy-lfgLfryRztU:-Nu4VbYy-lfgLfryRztV:br9epbv) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L133)):*
> ```Python
>     current_password = serializers.CharField(
>         write_only=True,
>         required=False,
> ```

see above comment required=False
___
*[`backend/api/serializers/user.py` line 142 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4VnMm3wrLYaPnF5UP:-Nu4VnMn3C5CEm7Yjr8x:b-on6x77) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L142)):*
> ```Python
>             "first_name": {
>                 "read_only": False,
>                 "required": False,
> ```

see above comment required=False. also don't need read_only=False
___
*[`backend/api/serializers/user.py` line 147 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4VpGJ8qCCtiYY2nYj:-Nu4VpGJ8qCCtiYY2nYk:b-on6x77) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L147)):*
> ```Python
>             "last_name": {
>                 "read_only": False,
>                 "required": False,
> ```

see above comment required=False. also don't need read_only=False
___
*[`backend/api/serializers/user.py` line 150 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4Vt-t7Kd6BeCLxPdc:-Nu4Vt-t7Kd6BeCLxPdd:b-on6x77) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L150)):*
> ```Python
>                 "min_length": 1,
>             },
>             "email": {"read_only": False, "required": False},
> ```

see above comment required=False. also don't need read_only=False
___
*[`backend/api/serializers/user.py` line 151 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4W5y76XULpz_86wCW:-Nu4W5y86zHgxJ852j1K:b-79ougd) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L151)):*
> ```Python
>             },
>             "email": {"read_only": False, "required": False},
>             "password": {"write_only": True, "required": False},
> ```

see above comment required=False.
___
*[`backend/api/tests/views/test_user.py` line 233 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4YbSA0q3MWwVPi617:-Nu4YbSA0q3MWwVPi618:b-20e2sm) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/tests/views/test_user.py#L233)):*
> ```Python
>         )
> 
>     def test_get_serializer_class__destroy(self):
> ```

this test is a mistake. destroy is the only default action that does not use a serializer. please delete this test.
___
*[`backend/api/tests/views/test_user.py` line 243 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4Ykp95UGdz3-fy7Au:-Nu4Ykp95UGdz3-fy7Av:b9yv9t) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/tests/views/test_user.py#L243)):*
> ```Python
>         """Retrieving a user uses the general serializer."""
>         self.assert_get_serializer_class(
>             serializer_class=UpdateUserSerializer,
> ```

from codeforlife.user.serializers import UserSerializer

serializer_class=UserSerializer (you may need to do UserSerializer[User])
___
*[`backend/api/tests/views/test_user.py` line 250 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4YxCtAyH-QYnAxZOH:-Nu4YxCuECSNZkZM2SvQ:b9yv9t) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/tests/views/test_user.py#L250)):*
> ```Python
>         """Listing users uses the general serializer."""
>         self.assert_get_serializer_class(
>             serializer_class=UpdateUserSerializer,
> ```

from codeforlife.user.serializers import UserSerializer

serializer_class=UserSerializer (you may need to do UserSerializer[User])
___
*[`backend/api/views/user.py` line 65 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4XdtL0OPs_bU2amQ5:-Nu4XdtL0OPs_bU2amQ6:b-z8sdif) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/views/user.py#L65)):*
> ```Python
>             return HandleIndependentUserJoinClassRequestSerializer
> 
>         return UpdateUserSerializer
> ```

replace this return with:

```py
if self.action == "partial_update":
    return UpdateUserSerializer

return super().get_serializer_class()
```

in other words, if the actions are "retrieve" or "list", the catch-all-return will return the read-only [user-serializer defined in the parent class](https://github.com/ocadotechnology/codeforlife-package-python/blob/fc40e3ac7c2b1d7b932e4bdd01a656846cc985a7/codeforlife/user/views/user.py#L17).

**Review by SKairinos — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317)* status: all files reviewed, 16 unresolved discussions (waiting on @faucomte97)

___
*[`backend/api/serializers/user.py` line 116 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4ZlSj7EUp3DevEqT2:-Nu4ZlSj7EUp3DevEqT3:bqo3w54) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L116)):*
> ```Python
> 
>     def create(self, validated_data):
>         add_to_newsletter = validated_data.pop("add_to_newsletter")
> ```

also need to pop date_of_birth. this is not need to create an independent user

**Review by faucomte97 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317)* status: all files reviewed, 16 unresolved discussions (waiting on @SKairinos)

___
*[`backend/Pipfile` line 48 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4QlqXCrRkfd6VA2BZ:-Nu4fDv689ltBp2b5dVQ:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/Pipfile#L48)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

not needed. [inherited from CFL package](https://github.com/ocadotechnology/codeforlife-package-python/blob/fc40e3ac7c2b1d7b932e4bdd01a656846cc985a7/Pipfile#L8)
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 182 at r1](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-NtzQUXPCyMlc_IdACyA:-Nu4d1K1E2uP7xU-iPiZ:b5m72vb) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/e6e6caf090fb0e2a8e0df192a78692daf24e8578/backend/api/serializers/user.py#L182)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

Because the data we pop from validated_data could be of any type. Therefore, `type(add_to_newsletter) == Any`. We need to hint to mypy that its a bool `add_to_newsletter: bool =  user_fields.pop("add_to_newsletter")`.

Yes, it should also be in CreateTeacherSerializer (that's my bad)
</blockquote></details>

Done. Also did it for DOB but not sure about the type.
___
*[`backend/api/serializers/user.py` line 97 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4UiMe7jbFn-FMiJoe:-Nu4_mznDiZ9MUnGzC6i:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L97)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

this is required when creating. replace with write_only=True as we do not need to return this value in the response.
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 116 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4ZlSj7EUp3DevEqT2:-Nu4_yb99yCZMMteDpov:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L116)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

also need to pop date_of_birth. this is not need to create an independent user
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 128 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4VRtAA4QG002STuGs:-Nu4aQld3R9Oo-jxkwC8:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L128)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

can remove required=False. Partial update forces everything to be required=False. No need to manually specify it.
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 133 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4VbYy-lfgLfryRztU:-Nu4aR7W9PEQT81_5qCa:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L133)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

see above comment required=False
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 142 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4VnMm3wrLYaPnF5UP:-Nu4a_tS1Cm8G84N7Iaf:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L142)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

see above comment required=False. also don't need read_only=False
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 147 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4VpGJ8qCCtiYY2nYj:-Nu4aa5dBtswIuvjSCwY:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L147)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

see above comment required=False. also don't need read_only=False
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 150 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4Vt-t7Kd6BeCLxPdc:-Nu4aabO4R8SyRw_GhbA:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L150)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

see above comment required=False. also don't need read_only=False
</blockquote></details>

Done.
___
*[`backend/api/serializers/user.py` line 151 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4W5y76XULpz_86wCW:-Nu4aaxi0GqXnP-NQ_62:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/serializers/user.py#L151)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

see above comment required=False.
</blockquote></details>

Done.
___
*[`backend/api/tests/views/test_user.py` line 233 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4YbSA0q3MWwVPi617:-Nu4eij310UZvpy4yqyC:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/tests/views/test_user.py#L233)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

this test is a mistake. destroy is the only default action that does not use a serializer. please delete this test.
</blockquote></details>

Done.
___
*[`backend/api/tests/views/test_user.py` line 243 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4Ykp95UGdz3-fy7Au:-Nu4eu4v9TlTt4FqixYV:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/tests/views/test_user.py#L243)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

from codeforlife.user.serializers import UserSerializer

serializer_class=UserSerializer (you may need to do UserSerializer[User])
</blockquote></details>

Done.
___
*[`backend/api/tests/views/test_user.py` line 250 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4YxCtAyH-QYnAxZOH:-Nu4eurADG1nqxdS0ajm:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/tests/views/test_user.py#L250)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

from codeforlife.user.serializers import UserSerializer

serializer_class=UserSerializer (you may need to do UserSerializer[User])
</blockquote></details>

Done.
___
*[`backend/api/views/user.py` line 65 at r2](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317#-Nu4XdtL0OPs_bU2amQ5:-Nu4f7zE-rWjLQz5rh6n:b-896fix) ([raw file](https://github.com/ocadotechnology/codeforlife-portal-react/blob/f64f0c5c5827bd3c5b15ee42c33f95f01cc9912a/backend/api/views/user.py#L65)):*
<details><summary><i>Previously, SKairinos (Stefan Kairinos) wrote…</i></summary><blockquote>

replace this return with:

```py
if self.action == "partial_update":
    return UpdateUserSerializer

return super().get_serializer_class()
```

in other words, if the actions are "retrieve" or "list", the catch-all-return will return the read-only [user-serializer defined in the parent class](https://github.com/ocadotechnology/codeforlife-package-python/blob/fc40e3ac7c2b1d7b932e4bdd01a656846cc985a7/codeforlife/user/views/user.py#L17).
</blockquote></details>

Done.

**Review by SKairinos — APPROVED** [review]
> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ocadotechnology/codeforlife-portal-react/317)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @faucomte97)
